File: app/src/main/java/com/example/fruitsdiary/usecase/addeditentry/AddEditEntryActivity.java
package com.example.fruitsdiary.usecase.addeditentry;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.example.fruitsdiary.R;
import com.example.fruitsdiary.model.Entry;
import com.example.fruitsdiary.model.FruitEntry;
import com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState;
import com.example.fruitsdiary.usecase.addeditentry.editfruit.EditFruitFragment;
import com.example.fruitsdiary.usecase.addeditentry.selectfruit.OnSelectFruitListener;
import com.example.fruitsdiary.usecase.addeditentry.selectfruit.SelectFruitFragment;
import com.example.fruitsdiary.util.DateUtils;

public class AddEditEntryActivity extends AppCompatActivity
        implements AddEditEntryFragment.OnAddEditListener,
        OnSelectFruitListener,
        EditFruitFragment.OnEditFruitListener {

    private AddEditEntryManager mAddEditEntryManager;
    private SelectFruitFragment mSelectFruitFragment;
    private EditFruitFragment mEditFruitFragment;
    private @EntryState
    int mEntryState;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_edit_entry);

        AddEditEntryIntent intent = new AddEditEntryIntent(getIntent());
        Entry entry = intent.getEntry();
        mEntryState = intent.getEntryState();

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);

        if (mEntryState == EntryState.CREATE) {
            actionBar.setTitle(DateUtils.getCurrentAppDate());
        } else {
            String date = DateUtils.convertServerDateToAppDate(entry.getDate());
            actionBar.setTitle(date);
        }


        AddEditEntryFragment addEditEntryFragment = new AddEditEntryFragment();
        mAddEditEntryManager = addEditEntryFragment;
        addEditEntryFragment.setOnAddEditListener(this);
        addEditEntryFragment.setOnSelectFruitListener(this);

        mEditFruitFragment = new EditFruitFragment();
        mEditFruitFragment.setOnEditFruitListener(this);

        mSelectFruitFragment = new SelectFruitFragment();
        mSelectFruitFragment.setSelectFruitListener(this);

        FragmentManager manager = getSupportFragmentManager();
        manager.beginTransaction()
                .add(R.id.fragment_container, addEditEntryFragment, AddEditEntryFragment.TAG)
                .commit();

    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        switch (mEntryState) {
            case EntryState.EDIT:
                inflater.inflate(R.menu.menu_edit_entry, menu);
                break;
            default:
            case EntryState.CREATE:
                inflater.inflate(R.menu.menu_create_entry, menu);
                break;
            case EntryState.VIEW:
                inflater.inflate(R.menu.menu_view_entry, menu);
                break;
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            case R.id.action_delete:
                mAddEditEntryManager.deleteEntry();
                return true;
            case R.id.action_save:
                mAddEditEntryManager.saveEntry();
                return true;
            default:
                return super.onOptionsItemSelected(item);

        }
    }

    private void setEntryStateToEdit() {
        if (mEntryState != EntryState.EDIT) {
            mEntryState = EntryState.EDIT;
            invalidateOptionsMenu();
        }
    }

    private void setEntryStateToView() {
        if (mEntryState != EntryState.VIEW) {
            mEntryState = EntryState.VIEW;
            invalidateOptionsMenu();
        }
    }

    @Override
    public void onBackPressed() {
        FragmentManager manager = getSupportFragmentManager();
        if (manager.findFragmentByTag(SelectFruitFragment.TAG) != null) {
            removeSelectFruitFragment(manager);
        } else if (manager.findFragmentByTag(EditFruitFragment.TAG) != null) {
            removeEditFruitFragment(manager);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public void onAddFruitEntryClick() {
        FragmentManager manager = getSupportFragmentManager();
        if (manager.findFragmentByTag(SelectFruitFragment.TAG) == null) {
            addSelectFruitFragment(manager);
        }
    }

    @Override
    public void onEntrySaved() {
        setEntryStateToView();
    }

    @Override
    public void onFruitSelected(@Nullable FruitEntry fruitEntry) {
        FragmentManager manager = getSupportFragmentManager();
        if (fruitEntry == null) {
            removeSelectFruitFragment(manager);
        } else {
            if (mAddEditEntryManager.contains(fruitEntry)) {
                fruitEntry = mAddEditEntryManager.getCorrespondingFruitEntry(fruitEntry);
            }
            mEditFruitFragment.setFruitEntry(fruitEntry);
            // remove fruit selection fragment to replace it with the fruit edition fragment
            manager.beginTransaction()
                    .setCustomAnimations(R.anim.slide_up, R.anim.slide_down)
                    .remove(mSelectFruitFragment)
                    .add(R.id.fragment_container, mEditFruitFragment, EditFruitFragment.TAG)
                    .commit();
        }
    }

    @Override
    public void onFruitEdited(@Nullable FruitEntry fruitEntry) {
        removeEditFruitFragment(getSupportFragmentManager());
        if (fruitEntry != null) {
            mAddEditEntryManager.addOrUpdateFruitEntry(fruitEntry);
            setEntryStateToEdit();
        }
    }

    @Override
    public void onFruitDeleted(@NonNull FruitEntry fruitEntry) {
        mAddEditEntryManager.deleteFruitEntry(fruitEntry);
        setEntryStateToEdit();
        removeEditFruitFragment(getSupportFragmentManager());
    }

    private void addSelectFruitFragment(FragmentManager manager) {
        manager.beginTransaction()
                .setCustomAnimations(R.anim.slide_up, 0)
                .add(R.id.fragment_container, mSelectFruitFragment, SelectFruitFragment.TAG)
                .commit();
        mAddEditEntryManager.showOverlay();
    }

    private void removeSelectFruitFragment(FragmentManager manager) {
        manager.beginTransaction()
                .setCustomAnimations(0, R.anim.slide_down)
                .remove(mSelectFruitFragment)
                .commit();
        mAddEditEntryManager.hideOverlay();
    }

    private void removeEditFruitFragment(FragmentManager manager) {
        manager.beginTransaction()
                .setCustomAnimations(0, R.anim.slide_down)
                .remove(mEditFruitFragment)
                .commit();
        mAddEditEntryManager.hideOverlay();
    }

}

File: app/src/main/java/com/example/fruitsdiary/usecase/addeditentry/AddEditEntryActivity.kt
package com.example.fruitsdiary.usecase.addeditentry

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentManager
import com.example.fruitsdiary.R
import com.example.fruitsdiary.model.FruitEntry
import com.example.fruitsdiary.usecase.addeditentry.AddEditEntryIntent.EntryState
import com.example.fruitsdiary.usecase.addeditentry.editfruit.EditFruitFragment
import com.example.fruitsdiary.usecase.addeditentry.selectfruit.OnSelectFruitListener
import com.example.fruitsdiary.usecase.addeditentry.selectfruit.SelectFruitFragment
import com.example.fruitsdiary.util.DateUtils

class AddEditEntryActivity : AppCompatActivity(), AddEditEntryFragment.OnAddEditListener, OnSelectFruitListener, EditFruitFragment.OnEditFruitListener {

    private lateinit var mAddEditEntryManager: AddEditEntryManager
    private lateinit var mSelectFruitFragment: SelectFruitFragment
    private lateinit var mEditFruitFragment: EditFruitFragment
    @EntryState
    private var mEntryState: Int = EntryState.VIEW


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_edit_entry)

        val intent = AddEditEntryIntent(intent)
        val entry  = intent.getEntry()
        mEntryState = intent.getEntryState()

        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)

        if (mEntryState == EntryState.CREATE) {
            actionBar?.title = DateUtils.getCurrentAppDate()
        } else {
            val date = DateUtils.convertServerDateToAppDate(entry.date)
            actionBar?.title = date
        }


        val addEditEntryFragment = AddEditEntryFragment()
        mAddEditEntryManager = addEditEntryFragment
        addEditEntryFragment.setOnAddEditListener(this)
        addEditEntryFragment.setOnSelectFruitListener(this)

        mEditFruitFragment = EditFruitFragment()
        mEditFruitFragment.setOnEditFruitListener(this)

        mSelectFruitFragment = SelectFruitFragment()
        mSelectFruitFragment.setSelectFruitListener(this)

        val manager = supportFragmentManager
        manager.beginTransaction()
                .add(R.id.fragment_container, addEditEntryFragment, AddEditEntryFragment.TAG)
                .commit()

    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        when (mEntryState) {
            EntryState.EDIT -> inflater.inflate(R.menu.menu_edit_entry, menu)
            EntryState.CREATE -> inflater.inflate(R.menu.menu_create_entry, menu)
            EntryState.VIEW -> inflater.inflate(R.menu.menu_view_entry, menu)
            else -> inflater.inflate(R.menu.menu_view_entry, menu)
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            R.id.action_delete -> {
                mAddEditEntryManager.deleteEntry()
                true
            }
            R.id.action_save -> {
                mAddEditEntryManager.saveEntry()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setEntryStateToEdit() {
        if (mEntryState != EntryState.EDIT) {
            mEntryState = EntryState.EDIT
            invalidateOptionsMenu()
        }
    }

    private fun setEntryStateToView() {
        if (mEntryState != EntryState.VIEW) {
            mEntryState = EntryState.VIEW
            invalidateOptionsMenu()
        }
    }

    override fun onBackPressed() {
        when {
            supportFragmentManager.findFragmentByTag(SelectFruitFragment.TAG) != null ->
                removeSelectFruitFragment(supportFragmentManager)
            supportFragmentManager.findFragmentByTag(EditFruitFragment.TAG) != null ->
                removeEditFruitFragment(supportFragmentManager)
            else -> super.onBackPressed()
        }
    }

    override fun onAddFruitEntryClick() {
        if (supportFragmentManager.findFragmentByTag(SelectFruitFragment.TAG) == null) {
            addSelectFruitFragment(supportFragmentManager)
        }
    }

    override fun onEntrySaved() {
        setEntryStateToView()
    }

    override fun onFruitSelected(originalFruitEntry: FruitEntry?) {
        var fruitEntry: FruitEntry? = null
        if (originalFruitEntry == null) {
            removeSelectFruitFragment(supportFragmentManager)
        } else {
            // if the fruit entry already exists, we will edit the existing one instead
            if (mAddEditEntryManager.contains(originalFruitEntry)) {
                fruitEntry = mAddEditEntryManager.getCorrespondingFruitEntry(originalFruitEntry)
            }
            mEditFruitFragment.setFruitEntry(fruitEntry ?: originalFruitEntry)
            // remove fruit selection fragment to replace it with the fruit edition fragment
            supportFragmentManager.beginTransaction()
                    .setCustomAnimations(R.anim.slide_up, R.anim.slide_down)
                    .remove(mSelectFruitFragment)
                    .add(R.id.fragment_container, mEditFruitFragment, EditFruitFragment.TAG)
                    .commit()
            mAddEditEntryManager.showOverlay()
        }
    }

    override fun onFruitEdited(fruitEntry: FruitEntry?) {
        removeEditFruitFragment(supportFragmentManager)
        if (fruitEntry != null) {
            mAddEditEntryManager.addOrUpdateFruitEntry(fruitEntry)
            setEntryStateToEdit()
        }
    }

    override fun onFruitDeleted(fruitEntry: FruitEntry) {
        mAddEditEntryManager.deleteFruitEntry(fruitEntry)
        setEntryStateToEdit()
        removeEditFruitFragment(supportFragmentManager)
    }

    companion object {
        private const val NO_ANIMATION = 0
    }

    private fun addSelectFruitFragment(manager: FragmentManager) {
        manager.beginTransaction()
                .setCustomAnimations(R.anim.slide_up, NO_ANIMATION)
                .add(R.id.fragment_container, mSelectFruitFragment, SelectFruitFragment.TAG)
                .commit()
        mAddEditEntryManager.showOverlay()
    }

    private fun removeSelectFruitFragment(manager: FragmentManager) {
        manager.beginTransaction()
                .setCustomAnimations(NO_ANIMATION, R.anim.slide_down)
                .remove(mSelectFruitFragment)
                .commit()
        mAddEditEntryManager.hideOverlay()
    }

    private fun removeEditFruitFragment(manager: FragmentManager) {
        manager.beginTransaction()
                .setCustomAnimations(NO_ANIMATION, R.anim.slide_down)
                .remove(mEditFruitFragment)
                .commit()
        mAddEditEntryManager.hideOverlay()
    }

}
